// gatsby-source-wordpress-experimental/steps/source-nodes/create-nodes/fetch-and-create-non-node-root-fields.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _store = _interopRequireDefault(require("../../../store"));

var _fetchGraphql = _interopRequireDefault(require("../../../utils/fetch-graphql"));

var _gatsbyCoreUtils = require("gatsby-core-utils");

var _formatLogMessage = require("../../../utils/format-log-message");

const fetchAndCreateNonNodeRootFields = async () => {
  const {
    remoteSchema: {
      nonNodeQuery
    },
    gatsbyApi: {
      helpers,
      pluginOptions
    }
  } = _store.default.getState();

  const activity = helpers.reporter.activityTimer((0, _formatLogMessage.formatLogMessage)(`fetch root fields`));
  activity.start();
  const {
    data
  } = await (0, _fetchGraphql.default)({
    query: nonNodeQuery,
    errorContext: `Error occured while fetching non-Node root fields.`
  });
  await helpers.actions.createNode(Object.assign({}, data, {
    id: `${pluginOptions.url}--rootfields`,
    parent: null,
    internal: {
      type: pluginOptions.schema.typePrefix,
      contentDigest: (0, _gatsbyCoreUtils.createContentDigest)(data)
    }
  }));
  activity.end();
};

var _default = fetchAndCreateNonNodeRootFields;
exports.default = _default;
