// gatsby-source-wordpress-experimental/steps/get-content-types.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.createContentTypeNodes = void 0;

require("source-map-support/register");

var _graphqlQueries = require("../utils/graphql-queries");

var _fetchGraphql = _interopRequireDefault(require("../utils/fetch-graphql"));

var _store = _interopRequireDefault(require("../store"));

var _helpers = require("./create-schema-customization/helpers");

var _formatLogMessage = require("../utils/format-log-message");

/**
 *
 * This func is temporary until support for this added in WPGQL
 * see https://github.com/wp-graphql/wp-graphql/issues/1045
 */
const createContentTypeNodes = async ({
  webhookBody: {
    preview
  },
  cache
}, pluginOptions) => {
  if (preview) {
    return null;
  }

  const contentTypesCacheKey = `WP_CONTENT_TYPES_DATA`;
  let cachedContentTypesData = await cache.get(contentTypesCacheKey);

  const state = _store.default.getState();

  const {
    schemaWasChanged
  } = state.remoteSchema;
  const refetchContentTypes = schemaWasChanged || !cachedContentTypesData;
  const {
    nodeQueries,
    fieldBlacklist
  } = state.remoteSchema;
  const {
    helpers
  } = state.gatsbyApi;
  const activity = helpers.reporter.activityTimer((0, _formatLogMessage.formatLogMessage)(`fetch content types`));

  if (refetchContentTypes) {
    if (pluginOptions.verbose) {
      activity.start();
    }

    const {
      data
    } = await (0, _fetchGraphql.default)({
      query: _graphqlQueries.availablePostTypesQuery
    });
    await cache.set(contentTypesCacheKey, data);
    cachedContentTypesData = data;
  }

  const contentTypes = cachedContentTypesData.postTypes.filter(contentType => !fieldBlacklist.includes(contentType.fieldNames.plural)).map(contentTypeObj => {
    const contentTypeQueryInfo = nodeQueries[contentTypeObj.fieldNames.plural];
    const {
      typeInfo
    } = contentTypeQueryInfo;
    return helpers.actions.createNode(Object.assign({}, typeInfo, {
      id: helpers.createNodeId(`${typeInfo.nodesTypeName}${typeInfo.singularName}--content-type`),
      parent: null,
      internal: {
        contentDigest: helpers.createContentDigest(typeInfo),
        type: (0, _helpers.buildTypeName)(`ContentType`)
      }
    }));
  });

  if (pluginOptions.verbose && refetchContentTypes) {
    activity.end();
  }

  return contentTypes;
};

exports.createContentTypeNodes = createContentTypeNodes;
